### PR TITLE
[ET-1232] Auto generate TicketsCommerce Checkout and Order Success page.

### PR DESCRIPTION
**Pull request by rafsuntaskin**
### 🎫 Ticket

[ET-1232] 
[ET-1233] 

### 🗒️ Description
* Auto generate TicketsCommerce Checkout page and Order Success page.

### 🎥 Artifacts 
🎥 screencast: https://d.pr/v/ehfkSe

### ✔️ Checklist
- [x] I've included a changelog entry in the readme.txt file. 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

[ET-1232]: https://theeventscalendar.atlassian.net/browse/ET-1232?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ET-1233]: https://theeventscalendar.atlassian.net/browse/ET-1233?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -181,6 +181,7 @@ Check out our extensive [knowledgebase](https://evnt.is/18wm) for articles on us
 
 = [TBD] TBD =
 
+* Enhancement - Auto generate TicketsCommerce Checkout page and Order success page. [ET-1232] [ET-1233]
```
Comment (juanfra) [inline readme.txt:184]:
```suggestion
* Enhancement - Auto generate checkout page when enabling Tickets Commerce. [ET-1232]
* Enhancement - Auto generate order success page when enabling Tickets Commerce. [ET-1233]

```

File: src/Tickets/Commerce/Payments_Tab.php [hunk 2]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+		if ( $page_exists ) {
+			return false;
+		}
+
+		$page_data = [
+			'post_status'    => 'publish',
+			'post_type'      => 'page',
+			'post_author'    => 1,
+			'post_name'      => $page_slug,
+			'post_title'     => 'Tickets Checkout',
```
Comment (juanfra) [inline src/Tickets/Commerce/Payments_Tab.php:162]: We may want to have that in gettext functions so that it also works in other languages (if translated).

**Review by juanfra — COMMENTED** [review]
Looking great! a couple comments

**Comment by bordoni** [comment]
Lets make sure the ticket of this PR is more descriptive.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 3]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+		$page_exists = get_page_by_path( $page_slug );
+
+		if ( $page_exists ) {
+			return false;
+		}
+
+		$page_data = [
+			'post_status'    => 'publish',
+			'post_type'      => 'page',
+			'post_author'    => 1,
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:160]: Post Author should be the current user.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 4]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+			'post_name'      => $page_slug,
+			'post_title'     => __( 'Tickets Checkout', 'event-tickets' ),
+			'post_content'   => '<!-- wp:shortcode -->[' . Checkout_Shortcode::get_wp_slug() . ']<!-- /wp:shortcode -->',
+			'post_parent'    => 0,
+			'comment_status' => 'closed',
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:165]: https://developer.wordpress.org/reference/functions/get_default_comment_status/

File: src/Tickets/Commerce/Payments_Tab.php [hunk 5]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	 *
+	 * @return bool
+	 */
+	public function maybe_auto_generate_checkout_page() {
+		if ( tribe( Checkout::class )->page_has_shortcode() ) {
+			return false;
+		}
+
+		$page_slug   = 'tickets-checkout';
+		$page_exists = get_page_by_path( $page_slug );
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:151]: We should use a meta to flag that it was auto-created.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 6]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	 *
+	 * @return bool
+	 */
+	public function maybe_auto_generate_order_success_page() {
+		if ( tribe( Success::class )->page_has_shortcode() ) {
+			return false;
+		}
+
+		$page_slug   = 'tickets-order';
+		$page_exists = get_page_by_path( $page_slug );
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:190]: Same as the order.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 7]
```diff
@@ -106,4 +117,99 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+			'post_content'   => '<!-- wp:shortcode -->[' . Success_Shortcode::get_wp_slug() . ']<!-- /wp:shortcode -->',
+			'post_parent'    => 0,
+			'comment_status' => 'closed',
+		];
+
+		$page_id = wp_insert_post( $page_data );
+
+		if ( is_wp_error( $page_id ) ) {
+			return false;
+		}
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:211]: Make this into a method so we only pass the title, slug and Shortcode.

File: src/Tickets/Hooks.php [hunk 8]
```diff
@@ -46,6 +46,7 @@ public function register() {
 	 */
 	protected function add_actions() {
 		add_action( 'tribe_settings_do_tabs', [ tribe( Payments_Tab::class ), 'register_tab' ], 15 );
+		add_action( 'tribe_settings_after_save_' . Payments_Tab::$slug, [ tribe( Payments_Tab::class ), 'maybe_generate_pages' ] );
```
Comment (bordoni) [inline src/Tickets/Hooks.php:49]: Dont hook directly create a bypass method in the Hooks class.

**Review by bordoni — CHANGES_REQUESTED** [review]

File: src/Tickets/Commerce/Payments_Tab.php (hunk 4 above)
Comment (rafsuntaskin) [reply src/Tickets/Commerce/Payments_Tab.php:165]: We should set this as `closed` as we should not allow having comments in Checkout or Success pages.

Comment (bordoni) [reply src/Tickets/Commerce/Payments_Tab.php:165]: Ahh ok. Good to know.

**Review by codingmusician — APPROVED** [review]

File: src/Tickets/Commerce/Payments_Tab.php [hunk 9]
```diff
@@ -14,6 +16,24 @@
...
+	public static $slug = 'payments';
+
+	/**
+	 * Option key prefix for page creation flag.
+	 *
+	 * @since TBD
+	 *
+	 * @var string
+	 */
+	public static $option_page_created = 'tec_tc_payments_page_created_';
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:35]: No need to make into a prefix, if the meta exists we consider that it was a automatically created page.

```suggestion
	public static $option_page_created_meta_key = 'tec_tc_payments_page_created';
```

File: src/Tickets/Commerce/Payments_Tab.php [hunk 10]
```diff
@@ -106,4 +126,131 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	public function create_page_with_shortcode( $page_slug, $page_name, $shortcode_name ) {
+		$page_data = [
+			'post_status'    => 'publish',
+			'post_type'      => 'page',
+			'post_author'    => get_current_user_id(),
+			'post_name'      => $page_slug,
+			'post_title'     => $page_name,
+			'post_content'   => '<!-- wp:shortcode -->[' . $shortcode_name . ']<!-- /wp:shortcode -->',
+			'post_parent'    => 0,
+			'comment_status' => 'closed',
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:227]:
```suggestion
			'comment_status' => 'closed',
			'meta_input' => [
				static::$option_page_create_meta_key => $shortcode_name
			],
```

File: src/Tickets/Commerce/Payments_Tab.php [hunk 11]
```diff
@@ -106,4 +126,131 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	/**
+	 * Set option to mark a page as created.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $page_slug URL slug of the page that needs to be flagged.
+	 */
+	public function flag_page_created( $page_slug ) {
+		tribe_update_option( static::$option_page_created . $page_slug, true );
+	}
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:242]: Remove these.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 12]
```diff
@@ -106,4 +126,131 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	 *
+	 * @since TBD
+	 *
+	 * @param string $page_slug URL slug of the page that needs to be checked.
+	 *
+	 * @return bool
+	 */
+	public function is_page_created( $page_slug ) {
+		return tribe_is_truthy( tribe_get_option( static::$option_page_created . $page_slug ) );
+	}
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:255]: Do a conditional for WP_Query Post using the meta key.

File: src/Tickets/Commerce/Payments_Tab.php [hunk 13]
```diff
@@ -106,4 +126,131 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+	 * @since TBD
+	 *
+	 * @param string $page_slug URL slug of the page.
+	 * @param string $page_name Name for page title.
+	 * @param string $shortcode_name Shortcode name that needs to be inserted in page content.
+	 *
+	 * @return int|\WP_Error
+	 */
+	public function create_page_with_shortcode( $page_slug, $page_name, $shortcode_name ) {
+		$page_data = [
```
Comment (bordoni) [inline src/Tickets/Commerce/Payments_Tab.php:219]: Check if the current user can create posts before creating.

**Review by bordoni — CHANGES_REQUESTED** [review]

**Review by codingmusician — APPROVED** [review]

File: src/Tickets/Commerce/Payments_Tab.php [hunk 14]
```diff
@@ -106,4 +126,135 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
 		 */
 		return apply_filters( 'tec_tickets_commerce_settings_top_level', $top_level_settings );
 	}
+
+	/**
+	 * Generate Chceckout and Success page if not found.
```
Comment (juanfra) [inline src/Tickets/Commerce/Payments_Tab.php:131]:
```suggestion
	 * Maybe Generate Checkout and Success page if not found.
```

File: src/Tickets/Commerce/Payments_Tab.php [hunk 15]
```diff
@@ -106,4 +126,135 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+		if ( ! tribe_is_truthy( $tc_enabled ) ) {
+			return;
+		}
+
+		$this->maybe_auto_generate_checkout_page();
+		$this->maybe_auto_generate_order_success_page();
+	}
+
+	/**
+	 * Generate Checkout page with shortcode if not available.
```
Comment (juanfra) [inline src/Tickets/Commerce/Payments_Tab.php:148]:
```suggestion
	 * Generate Checkout page with the shortcode if the page is non-existent.
```

File: src/Tickets/Commerce/Payments_Tab.php [hunk 16]
```diff
@@ -106,4 +126,135 @@ class="tec-tickets__admin-settings-tickets-commerce-toggle-checkbox tribe-depend
...
+
+		if ( is_wp_error( $page_id ) ) {
+			return false;
+		}
+
+		return tribe_update_option( Settings::$option_checkout_page, $page_id );
+	}
+
+	/**
+	 * Generate Order Success page with shortcode if not available.
```
Comment (juanfra) [inline src/Tickets/Commerce/Payments_Tab.php:177]:
```suggestion
	 * Generate Order Success page with the shortcode if the page is non-existent.
```

**Review by juanfra — APPROVED** [review]

**Review by juanfra — APPROVED** [review]

**Pull request merged**